LightGBM/kaggle.py

```diff
@@ -0,0 +1,110 @@
+import pandas as pd
+
+import lightgbm as lgb
+
+from datetime import datetime, timedelta
+from tqdm import tqdm
+
+from data import data_frames, optimize_df, melt_and_merge, features, lgb_dataset
+import model
+
+
+# Global constants
+FIRST_DATE = datetime(2016, 4, 25) #  first date of validation and evaluation days
+MAX_LAG = timedelta(days=57)
+
+
+def next_day_features(df, forecast_date):
+    """
+    Create features of the next day to forecast.
+
+    Args:
+        df            = [pd.DataFrame] long format dataframe
+        forecast_date = [datetime] forecast date
+
+    Returns [pd.DataFrame]:
+        Dataframe with features for the next day to forecast.
+    """
+    forecast_df = df[ (df['date'] >= forecast_date - MAX_LAG) & (df['date'] <= forecast_date) ].copy()    
+    forecast_df = features(forecast_df, submission=True)   
+    return forecast_df
+
+
+def make_submission(df):
+    """
+    Create dataframe in the correct format for submission.
+
+    Args:
+        df = [pd.DataFrame] pandas dataframe
+
+    Returns [pd.DataFrame]:
+        Submission dataframe.
+    """
+    cols = [f"F{i}" for i in range(1, 29)] 
+
+    submission = df.loc[df['date'] >= FIRST_DATE, ['id', 'sales']].copy()
+    submission['F'] = [f'F{rank}' for rank in submission.groupby('id')['id'].cumcount() + 1]
+    submission = submission.set_index(['id', 'F']).unstack()['sales'][cols].reset_index()
+    submission.fillna(0., inplace=True)
+    submission.sort_values("id", inplace=True)
+    submission.reset_index(drop=True, inplace=True)
+
+    # make a dummy evaluation forecast
+    submission_eval = submission.copy()
+    submission_eval['id'] = submission_eval['id'].str.replace('validation', 'evaluation')
+    submission = pd.concat([submission, submission_eval], axis=0, sort=False)    
+
+    return submission
+
+
+def infer(model, calendar, prices, sales):
+    """
+    Infer the unit sales with the model.
+
+    Args:
+        model    = [lgb.Booster] trained LightGBM model
+        calendar = [pd.DataFrame] dates of product sales
+        prices   = [pd.DataFrame] price of the products sold per store and date
+        sales    = [pd.DataFrame] historical daily unit sales data per product and store 
+
+    Returns [pd.DataFrame]:
+        Submission dataframe.
+    """
+    # create test dataset for submission
+    df = melt_and_merge(calendar, prices, sales, submission=True)
+
+    # forecast the 28 days for validation
+    for day in tqdm(range(0, 28)):
```

Comment:
Do you only make a submission for the validation data?

Reply:
Yes indeed, the source kaggle notebook does this, but I also thought about whether to change that to also include evaluation data (which is now just a copy of validation and concatenated to it for the submission).  For the evaluation data, we probably eventually want to use the ground truth when the competition releases it, so I thought it was fine to leave it like this, and it saves some time to create the submssion (30+ min for only validation data, so probably would be 1 hour if we include evaluation data).

Reply:
Have you run the code on a GCP VM? It might be interesting to compare test and train times and we would need to run all the code on the same platform for fair comparisons.